Make fieldAbsorberTest a test for the CI (#4682)
ComputationalRadiationPhysics/picongpu: share/picongpu/tests/FieldAbsorber/README.rst

.. _tests-fieldAbsorber:

Current Source Radiating in an Unbounded Three-Dimensional Region
=================================================================

.. sectionauthor:: Klaus Steiniger, Sergey Ermakov
.. moduleauthor:: Sergei Bastrakov, Klaus Steiniger

Example of a rectangular conductor with a steady current.

This setup is based on [Taflove2005]_, section 7.11.1.
The difference is we consider both 2D and 3D cases, and grid size may be increased due to our absorber being part of the simulation area, not located outside of it as in the book.

> The conductor is located within the center of the x-y-plane, oriented along the z-axis and infinitely extended along this axis, but current flows along y.

Reviewer: Question: How can I have a wire in z that transports current in y?

Author: Well, the orientation of the current density vector just depends on which faces of the square shaped wire you apply the potential difference ;-)

Honestly, this is just the Taflove setup. I do not know exactly how this choice came about. I wonder, whether there is a correlation between the fact that there is no radiation along the x axis in this case, and the choice of measurement points.

Reviewer: Okay

The conductor properties can be adjusted in `include/picongpu/param/fieldBackground.param` within the `class FieldBackgroundJ`.
Specifically, the variable `halfWidth` adjusts its edge length and can be used to apply the test with meaningful results to higher-order solvers, too.
The current in the wire ramps up over time according to a differentiated Gaussian.
This defines the current density amplitude, too.
Therefore, the total current through the wire scales with the wire's halfWidth.

The simulation is performed twice.
First the actual PML test is performed with a small simulation volume, here 60x60x60 cells.
The simulation runs for 1000 steps which is much longer than the steady state response.

> Second, a reference simulation is performed with a large simulation volume of 1060x1060x1060 cells.

Reviewer: `1060` is not dividable by `8`, thus this does not fit into standard supercells and will be auto-adjusted.

Reviewer: This is also not in agreement with the actual call. There `660` cells are used.

Reviewer: while discussing offline, you @steindev mentioned, that you adjusted the supercell size. However, github shows me that no cfg (except for background field) file has been altered.

Author: Because `memory.param` has not been changed. But it defines `using SuperCellSize = typename mCT::shrinkTo<mCT::Int<2, 2, 4>, simDim>::type;`

> PML boundaries are applied at the boundaries perpendicular to the x and y axes.

Reviewer: ```suggestion
PML boundaries are applied at the boundaries with normals in the x and y axes.
```

Author: Done

> Periodic boundaries are applied at the boundary perpendicular to the z axis.

Reviewer: ```suggestion
Periodic boundaries are applied at the boundary with a normal in the z axis.
```

Author: Done

Each component of the electric field is probed over time at two different positions, each of which has the same distance from the conductor in the two simulations.
That is, these points have to be less than 30 cells away from the conductor.

Since the reference grid is sufficiently large to avoid reflections from the boundary at the points of measurement, the PML quality is defined as the relative error

.. math::

  \mathrm{Rel. error}\rvert^{n}_{i,j} =
    \frac{\left| E\rvert^{n}_{i,j} - E_\mathrm{ref}\rvert^{n}_{i,j}  \right|}{\left| E_\mathrm{ref,max}\rvert_{i,j}  \right|}

with :math:`n` being the time step, :math:`(i,j)` the distance in cells from the conductor, :math:`E_\mathrm{ref,max}\rvert_{i,j}` the maximum amplitude of the reference field at :math:`(i,j)` over the total simulation duration.

> According to [Taflove2005]_, the PML quality should be of the order of :math:`10^{-5}` for a 10 cell CPML.

Reviewer: What is a **C**PML?

Author: Convolutional PML. There is also Uniaxial PML (UPML), all defined in Taflove.

Reviewer: Okay
